### PR TITLE
Gh 1375 graphlibrary properties and schema ids overridden

### PR DESCRIPTION
**Pull request by m55624**

### CODE REVIEW

File: store-implementation/federated-store/src/test/java/uk/gov/gchq/gaffer/federatedstore/FederatedStoreGraphVisibilityTest.java [hunk 1]
```diff
@@ -54,7 +53,6 @@
 
     @Before
     public void setUp() throws Exception {
-        HashMapGraphLibrary.clear();
```
Comment (GCHQDeveloper404) [inline store-implementation/federated-store/src/test/java/uk/gov/gchq/gaffer/federatedstore/FederatedStoreGraphVisibilityTest.java]: Why are you removing the clearing of the graph before it is used?

File: store-implementation/federated-store/src/test/java/uk/gov/gchq/gaffer/federatedstore/FederatedStoreSchemaTest.java [hunk 2]
```diff
@@ -56,33 +56,28 @@
...
         fStore.initialise(TEST_FED_STORE, null, FEDERATED_PROPERTIES);
 
-        HashMapGraphLibrary.clear();
-    }
-
-    @After
-    public void tearDown() throws Exception {
-        HashMapGraphLibrary.clear();
-
+        library.clear();
```
Comment (GCHQDeveloper404) [inline store-implementation/federated-store/src/test/java/uk/gov/gchq/gaffer/federatedstore/FederatedStoreSchemaTest.java]: 👍

File: store-implementation/federated-store/src/test/java/uk/gov/gchq/gaffer/federatedstore/FederatedStoreSchemaTest.java [hunk 3]
```diff
@@ -56,33 +56,28 @@
 
     @Before
     public void setUp() throws Exception {
-        ACCUMULO_PROPERTIES.setStoreClass(SingleUseMockAccumuloStore.class);
+        ACCUMULO_PROPERTIES.setId("accProp");
+        ACCUMULO_PROPERTIES.setStoreClass(MockAccumuloStore.class.getName());
```
Comment (GCHQDeveloper404) [inline store-implementation/federated-store/src/test/java/uk/gov/gchq/gaffer/federatedstore/FederatedStoreSchemaTest.java]: remove ".getName()"

File: core/graph/src/test/java/uk/gov/gchq/gaffer/graph/GraphTest.java [hunk 4]
```diff
@@ -1203,7 +1203,6 @@ public void shouldThrowExceptionIfGraphIdIsInvalid() throws Exception {
     @Test
     public void shouldBuildGraphUsingGraphIdAndLookupSchema() throws Exception {
         // Given
-        HashMapGraphLibrary.clear();
```
Comment (GCHQDeveloper404) [inline core/graph/src/test/java/uk/gov/gchq/gaffer/graph/GraphTest.java]: Why are you removing this, before it is used. look @ line: 1266 👀

File: core/store/src/main/java/uk/gov/gchq/gaffer/store/library/GraphLibrary.java [hunk 5]
```diff
@@ -216,4 +221,24 @@ private void checkExisting(final String graphId, final byte[] schema, final Stor
             }
         }
     }
+
+    private void checkSchemaExists(final Schema schema) {
+        if (null != getSchema(schema.getId())) {
+            if (!JsonUtil.equals(getSchema(schema.getId()).toJson(false), schema.toJson(false))) {
```
Comment (GCHQDeveloper404) [inline core/store/src/main/java/uk/gov/gchq/gaffer/store/library/GraphLibrary.java]: Maybe you can use `JSONSerialiser.serialise()`

File: core/store/src/main/java/uk/gov/gchq/gaffer/store/library/GraphLibrary.java [hunk 6]
```diff
@@ -216,4 +221,24 @@ private void checkExisting(final String graphId, final byte[] schema, final Stor
...
+                throw new OverwritingException("schemaId " + schema.getId() + " already exists with a different schema:\n"
+                        + "existing schema:\n" + StringUtil.toString(getSchema(schema.getId()).toJson(false))
+                        + "\nnew schema:\n" + StringUtil.toString(schema.toJson(false)));
+            }
+        }
+    }
+
+    private void checkPropertiesExist(final StoreProperties properties) {
+        if (null != getProperties(properties.getId())) {
+            if (!getProperties(properties.getId()).equals(properties)) {
```
Comment (GCHQDeveloper404) [inline core/store/src/main/java/uk/gov/gchq/gaffer/store/library/GraphLibrary.java]: I "think" you need to compare the properties within StoreProperties.

File: core/store/src/test/java/uk/gov/gchq/gaffer/store/library/AbstractGraphLibraryTest.java [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+        final Schema unknownSchema = graphLibrary.getSchema(TEST_UNKNOWN_SCHEMA_ID);
+
+        // Then
+        assertNull(unknownSchema);
+    }
+
+    @Test
+    public void shouldThrowExceptionWhenNewStorePropertiesAreAddedWithSamePropertiesIdAndDifferentProperties() {
+        // Given
+        final StoreProperties tempStoreProperties = new StoreProperties(TEST_PROPERTIES_ID);
```
Comment (GCHQDeveloper404) [inline core/store/src/test/java/uk/gov/gchq/gaffer/store/library/AbstractGraphLibraryTest.java]: `        final StoreProperties tempStoreProperties = StoreProperties.loadStoreProperties(storeProperties.getProperties());`

**Comment by GCHQDeveloper404** [comment]
Is there a test for if not can you add

```
   // Given
        final Schema tempSchema = new Schema.Builder()
                .merge(schema)
                .id(TEST_SCHEMA_ID)
                .build();

        // When
        graphLibrary.addSchema(schema);

        // Then
        graphLibrary.addSchema(tempSchema);
        
       assertTrue( library is same, not error thrown);
```

File: core/store/src/test/java/uk/gov/gchq/gaffer/store/library/AbstractGraphLibraryTest.java [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+
+        // Then
+        assertEquals(null, graphLibrary.getIds(TEST_GRAPH_ID));
+        assertEquals(null, graphLibrary.getSchema(TEST_SCHEMA_ID));
+        assertEquals(null, graphLibrary.getProperties(TEST_PROPERTIES_ID));
+
+    }
+
+    @Test
+    public void shouldAddAndGetMultipleIdsInGraphLibrary() {
```
Comment (GCHQDeveloper404) [inline core/store/src/test/java/uk/gov/gchq/gaffer/store/library/AbstractGraphLibraryTest.java]: You haven't "Got" in this test.

File: core/store/src/test/java/uk/gov/gchq/gaffer/store/library/AbstractGraphLibraryTest.java [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+        // When / Then
+        try {
+            graphLibrary.addProperties(null);
+        } catch (final IllegalArgumentException e) {
+            assertTrue(e.getMessage().contains("Id is invalid: null"));
+        }
+    }
+
+    @Test
+    public void shouldThrowExceptionWhenGraphIdWithDifferentPropertiesExists() {
```
Comment (GCHQDeveloper404) [inline core/store/src/test/java/uk/gov/gchq/gaffer/store/library/AbstractGraphLibraryTest.java]: add won for when schema and properties is the same. what is the expected behaviour

File: core/store/src/test/java/uk/gov/gchq/gaffer/store/library/AbstractGraphLibraryTest.java [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+        try {
+            graphLibrary.addProperties(tempStoreProperties);
+            fail("Exception expected");
+        } catch (final OverwritingException e) {
+            assertTrue(e.getMessage().contains("already exists with a different store properties"));
+        }
+    }
+
+    @Test
+    public void shouldThrowExceptionWhenNewSchemaIsAddedWithSameSchemaIdAndSchema() {
```
Comment (GCHQDeveloper404) [inline core/store/src/test/java/uk/gov/gchq/gaffer/store/library/AbstractGraphLibraryTest.java]: I think this test is wrong.
1) The schema is not the same for the SchemId in library, so it doesn't match the title of this test, "WithSameSchemaIdAndSchema".

2) If you add a new schema and it is the same as the existing schema an error shouldn't be thrown. I think is should possibly be silently ignore.

File: core/store/src/main/java/uk/gov/gchq/gaffer/store/library/GraphLibrary.java (hunk 5 above)
Comment (m55624) [reply core/store/src/main/java/uk/gov/gchq/gaffer/store/library/GraphLibrary.java]: .toJson() uses that method

File: core/store/src/test/java/uk/gov/gchq/gaffer/store/library/AbstractGraphLibraryTest.java (hunk 9 above)
Comment (m55624) [reply core/store/src/test/java/uk/gov/gchq/gaffer/store/library/AbstractGraphLibraryTest.java]: This is already there

File: core/store/src/test/java/uk/gov/gchq/gaffer/store/library/AbstractGraphLibraryTest.java (hunk 10 above)
Comment (m55624) [reply core/store/src/test/java/uk/gov/gchq/gaffer/store/library/AbstractGraphLibraryTest.java]: Ah, this was just meant to say different schema.  It is correct for the exception to be thrown when adding a second Schema with the same ID but different Schema.

File: core/store/src/main/java/uk/gov/gchq/gaffer/store/library/GraphLibrary.java (hunk 6 above)
Comment (m55624) [reply core/store/src/main/java/uk/gov/gchq/gaffer/store/library/GraphLibrary.java]: I have updated this logic in GraphLibrary now

File: core/graph/src/test/java/uk/gov/gchq/gaffer/graph/GraphTest.java (hunk 4 above)
Comment (m55624) [reply core/graph/src/test/java/uk/gov/gchq/gaffer/graph/GraphTest.java]: It is no longer a static method.  a graphLibrary would have to be instantiated then cleared each time, which currently it doesn't do.  It works without clearing it.

Comment (GCHQDeveloper404) [reply core/graph/src/test/java/uk/gov/gchq/gaffer/graph/GraphTest.java]: 👀 HashMapGraphLibrary lines: 30-32.

Comment (GCHQDeveloper404) [reply core/graph/src/test/java/uk/gov/gchq/gaffer/graph/GraphTest.java]: @m55624  ?

**Pull request merged**